openfe/protocols/openmm_utils/charge_generation.py

```diff
@@ -0,0 +1,415 @@
+        )
+
+
+def assign_offmol_partial_charges(
+    offmol: OFFMol,
+    overwrite: bool,
+    method: Literal['am1bcc', 'am1bccelf10', 'nagl', 'espaloma'],
+    toolkit_backend: Literal['ambertools', 'openeye', 'rdkit'],
+    generate_n_conformers: Optional[int],
+    nagl_model: Optional[str],
+) -> None:
+    """
+    Assign partial charges to an OpenFF Molecule based on a selected method.
+
+    Parameters
+    ----------
+    offmol : openff.toolkit.Molecule
+      The Molecule to assign partial charges to.
+    overwrite : bool
+      Whether or not to overwrite any existing non-zero partial charges.
+      Note that zeroed charges will always be overwritten.
+    method : Literal['am1bcc', 'am1bccelf10', 'nagl', 'espaloma']
+      Partial charge assignement method.
+      Supported methods include; am1bcc, am1bccelf10, nagl, and espaloma.
+    toolkit_backend : Literal['ambertools', 'openeye', 'rdkit']
+      OpenFF toolkit backend employed for charge generation.
+      Supported options:
+        * ``ambertools``: selects both the AmberTools and RDKit Toolkit Wrapper
+        * ``openeye``: selects the OpenEye toolkit Wrapper
+        * ``rdkit``: selects the RDKit toolkit Wrapper
+      Note that the ``rdkit`` backend cannot be used for `am1bcc` or
+      ``am1bccelf10`` partial charge methods.
+    generate_n_conformers : Optional[int]
+      Number of conformers to generate for partial charge generation.
+      If ``None`` (default), the input conformer will be used.
+      Values greater than 1 can only be used alongside ``am1bccelf10``.
+    nagl_model : Optional[str]
+      The NAGL model to use for charge assignment if method is ``nagl``.
+      If ``None``, the latest am1bcc NAGL charge model is used.
+
+    Raises
+    ------
+    ValueError
+      If the ``toolkit_backend`` is not suported by the selected ``method``.
+      If ``generate_n_conformers`` is ``None``, but the input ``offmol``
+      has no associated conformers.
+      If the number of conformers passed or generated exceeds the number
+      of conformers selected by the partial charge ``method``.
+    """
+
+    # If you have non-zero charges and not overwriting, just return
+    if (offmol.partial_charges is not None and np.any(offmol.partial_charges)):
+        if not overwrite:
+            return
+
+    # Dictionary for each available charge method
+    # The idea of this pattern is to allow for maximum flexibility by
+    # allowing for swapping out method calls as necessary.
+    #
+    # Must include:
+    # 1. `confgen_func`: the conformer generation method
+    # 2. `charge_func`: the partial charge assignment method
+    # 2. `backends`: the allowed backends for the method
+    # 3. `max_conf`: maximum number of allowed conformations for the method
+    # 4. `charge_extra_kwargs`: any additional kwargs to be passed to the
+    #    partial charge assignment method beyond the input offmol and
+    #    the toolkitregistry
+    CHARGE_METHODS = {
```

Comment:
The use case I'm imagining here is that someone can re-use most of the tooling here and just add more charge methods as they go along.

I.e. you can just redefine this method and add a "random conformer am1bcc" that's rdkit backend only and then the main thing is that it uses a different confgen_func.

Not sure if that's overengineering though.